spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkParquetReader.java

```diff
@@ -248,4 +262,66 @@ public void testMissingRequiredWithoutDefault() {
         .isInstanceOf(IllegalArgumentException.class)
         .hasMessage("Missing required field: missing_str");
   }
+
+  @Test
+  public void testUnshreddedVariants() throws IOException {
+    Schema schema = new Schema(required(1, "v", Types.VariantType.get()));
+    StructType sparkSchema =
+        new StructType(
+            new StructField[] {
+              new StructField("v", DataTypes.VariantType, true, Metadata.empty())
+            });
+    List<InternalRow> rows = Lists.newArrayList(RandomData.generateSpark(schema, 10, 0L));
+    List<Record> tableRecords = writeAndRead(schema, sparkSchema, rows);
+
+    assertThat(tableRecords).hasSameSizeAs(rows);
+    for (int i = 0; i < tableRecords.size(); i++) {
+      GenericsHelpers.assertEqualsUnsafe(schema.asStruct(), tableRecords.get(i), rows.get(i));
+    }
+  }
+
+  private static List<Record> writeAndRead(
```

Comment:
I don't think that this needs a new `writeAndRead` method. The `GenericParquetWriter` used by the shared `writeAndRead` method supports variants.